// userscript-metadata/dist/common.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.id = id;
exports.not = not;
exports.fromMaybeUndefined = fromMaybeUndefined;
exports.isEmpty = isEmpty;
exports.containsWhitespace = containsWhitespace;
exports.containsOnlyWhitespace = containsOnlyWhitespace;
exports.containsLeadingWhitespace = containsLeadingWhitespace;
exports.containsTrailingWhitespace = containsTrailingWhitespace;
exports.containsLineBreaks = containsLineBreaks;
exports.replaceHyphens = replaceHyphens;
exports.replaceUnderscores = replaceUnderscores;
const r = exports.r = String.raw; // Python-style raw strings
function id(x) {
    return x;
}
function not(f) {
    return x => !f(x);
}
function fromMaybeUndefined(fallback, x) {
    return x === undefined ? fallback : x;
}
function isEmpty(x) {
    return x === "";
}
function containsWhitespace(x) {
    return (/\s/.test(x)
    );
}
function containsOnlyWhitespace(x) {
    return (/^\s*$/.test(x)
    );
}
function containsLeadingWhitespace(x) {
    return (/^\s/.test(x)
    );
}
function containsTrailingWhitespace(x) {
    return (/\s$/.test(x)
    );
}
function containsLineBreaks(x) {
    return x.includes("\n");
}
function replaceHyphens(s) {
    return s.replace(/\-/g, "_");
}
function replaceUnderscores(s) {
    return s.replace(/_/g, "-");
}
