// userscript-metadata/dist/data.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.DEFAULT_WARNINGS = exports.DEFAULT_ITEMS = exports.RUN_AT_VALUES = exports.DOCUMENT_START = exports.DOCUMENT_IDLE = exports.DOCUMENT_END = undefined;

var _connect = require("./connect");

var _item = require("./item");

var _patterns = require("./patterns");

var _key = require("./key");

var _messages = require("./messages");

var Msg = _interopRequireWildcard(_messages);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

const DOCUMENT_END = exports.DOCUMENT_END = "document-end";
const DOCUMENT_IDLE = exports.DOCUMENT_IDLE = "document-idle";
const DOCUMENT_START = exports.DOCUMENT_START = "document-start";
const RUN_AT_VALUES = exports.RUN_AT_VALUES = [DOCUMENT_END, DOCUMENT_IDLE, DOCUMENT_START];
const INJECT_INTO_VALUES = ["auto", "content", "page"];
const SEMVER_REGEX = /^(\d+)(?:\.(\d+))*$/;
const DEFAULT_ITEMS = exports.DEFAULT_ITEMS = {
    author: new _item.StringItem({
        key: _key.KEY.author,
        unique: true,
        required: false
    }),
    connect: new _item.StringItem({
        key: _key.KEY.connect,
        unique: false,
        required: false,
        constraints: [{
            requirement: _connect.isValidConnectValue,
            message: _connect.validConnectValueRequired
        }]
    }),
    date: new _item.StringItem({
        key: _key.KEY.date,
        unique: true,
        required: false
    }),
    description: new _item.StringItem({
        key: _key.KEY.description,
        unique: true,
        required: false
    }),
    downloadURL: new _item.StringItem({
        key: _key.KEY.downloadURL,
        unique: true,
        required: false
    }),
    exclude: new _item.StringItem({
        key: _key.KEY.exclude,
        unique: false,
        required: false
    }),
    grant: new _item.StringItem({
        key: _key.KEY.grant,
        unique: false,
        required: false
    }),
    homepageURL: new _item.StringItem({
        key: _key.KEY.homepageURL,
        unique: true,
        required: false
    }),
    icon: new _item.StringItem({
        key: _key.KEY.icon,
        unique: true,
        required: false
    }),
    include: new _item.StringItem({
        key: _key.KEY.include,
        unique: false,
        required: false
    }),
    inject_into: new _item.StringItem({
        key: _key.KEY.inject_into,
        unique: true,
        required: false,
        constraints: [{
            requirement: v => INJECT_INTO_VALUES.includes(v),
            message: `Value must be one of ${INJECT_INTO_VALUES.map(x => `"${x}"`).join(", ")}.`
        }]
    }),
    license: new _item.StringItem({
        key: _key.KEY.license,
        unique: true,
        required: false
    }),
    match: new _item.StringItem({
        key: _key.KEY.match,
        unique: false,
        required: false,
        constraints: [{
            requirement: _patterns.isMatchPattern,
            message: Msg.validMatchPatternRequired
        }]
    }),
    name: new _item.StringItem({
        key: _key.KEY.name,
        unique: true,
        required: true
    }),
    namespace: new _item.StringItem({
        key: _key.KEY.namespace,
        unique: true,
        required: false
    }),
    noframes: new _item.BooleanItem({
        key: _key.KEY.noframes
    }),
    require: new _item.StringItem({
        key: _key.KEY.require,
        unique: false,
        required: false
    }),
    resource: new _item.StringItem({
        key: _key.KEY.resource,
        unique: false,
        required: false
    }),
    run_at: new _item.StringItem({
        key: _key.KEY.run_at,
        unique: true,
        required: false,
        constraints: [{
            requirement: v => RUN_AT_VALUES.includes(v),
            message: `Value must be one of ${RUN_AT_VALUES.map(x => `"${x}"`).join(", ")}.`
        }]
    }),
    updateURL: new _item.StringItem({
        key: _key.KEY.updateURL,
        unique: true,
        required: false
    }),
    version: new _item.StringItem({
        key: _key.KEY.version,
        unique: true,
        required: false,
        constraints: [{
            requirement: v => SEMVER_REGEX.test(v),
            message: Msg.semanticVersionRequired
        }]
    })
};
const DEFAULT_WARNINGS = exports.DEFAULT_WARNINGS = [entries => !entries.some(entry => entry.key === _key.KEY.match || entry.key === _key.KEY.include) ? [{
    summary: Msg.noMatchOrIncludeSummary,
    description: Msg.noMatchOrIncludeDescription
}] : [], entries => entries.some(entry => entry.key === _key.KEY.include && !(0, _patterns.isIncludePattern_regex)(entry.value)) ? [{
    summary: Msg.matchInsteadOfIncludeSummary,
    description: Msg.matchInsteadOfIncludeDescription
}] : []];
