// userscript-metadata/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _data = require("./data");

Object.keys(_data).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _data[key];
    }
  });
});

var _item = require("./item");

Object.keys(_item).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _item[key];
    }
  });
});

var _main = require("./main");

Object.keys(_main).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _main[key];
    }
  });
});

var _patterns = require("./patterns");

Object.keys(_patterns).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _patterns[key];
    }
  });
});

var _syntax = require("./syntax");

Object.keys(_syntax).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _syntax[key];
    }
  });
});

var _types = require("./types");

Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});

var _key = require("./key");

Object.defineProperty(exports, "KEY", {
  enumerable: true,
  get: function () {
    return _key.KEY;
  }
});

var _read = require("./read");

Object.defineProperty(exports, "extractBlock", {
  enumerable: true,
  get: function () {
    return _read.extractBlock;
  }
});
Object.defineProperty(exports, "parseBlock", {
  enumerable: true,
  get: function () {
    return _read.parseBlock;
  }
});

var _stringify = require("./stringify");

Object.defineProperty(exports, "stringify", {
  enumerable: true,
  get: function () {
    return _stringify.stringify;
  }
});

var _validation = require("./validation");

Object.defineProperty(exports, "validate", {
  enumerable: true,
  get: function () {
    return _validation.validate;
  }
});
Object.defineProperty(exports, "validateWith", {
  enumerable: true,
  get: function () {
    return _validation.validateWith;
  }
});
