// userscript-metadata/dist/item.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.StringItem = exports.NonBooleanItem = exports.BooleanItem = exports.BaseItem = undefined;

var _types = require("./types");

var _common = require("./common");

var _messages = require("./messages");

var Msg = _interopRequireWildcard(_messages);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

class BaseItem {
    constructor(description) {
        this.description = description;
        this.key = description.key;
    }
}
exports.BaseItem = BaseItem;
class BooleanItem extends BaseItem {
    constructor() {
        super(...arguments);
        this.unique = true;
        this.required = false;
    }
    validate(value) {
        return value === true ? (0, _types.Right)(true) : (0, _types.Left)(Msg.onlyTrueAllowed);
    }
}
exports.BooleanItem = BooleanItem;
class NonBooleanItem extends BaseItem {
    constructor(description) {
        super(description);
        this.description = description;
        this.required = description.required;
        this.unique = description.unique;
        this.constraints = (0, _common.fromMaybeUndefined)([], description.constraints);
    }
}
exports.NonBooleanItem = NonBooleanItem;
class StringItem extends NonBooleanItem {
    constructor(description) {
        super(description);
        this.description = description;
        this.constraints = [{
            requirement: (0, _common.not)(_common.containsOnlyWhitespace),
            message: Msg.someNonWhitespaceRequired
        }, {
            requirement: (0, _common.not)(_common.containsLeadingWhitespace),
            message: Msg.leadingWhitespaceNotAllowed
        }, {
            requirement: (0, _common.not)(_common.containsTrailingWhitespace),
            message: Msg.trailingWhitespaceNotAllowed
        }, {
            requirement: (0, _common.not)(_common.containsLineBreaks),
            message: Msg.lineBreaksNotAllowed
        }].concat((0, _common.fromMaybeUndefined)([], description.constraints));
    }
    butRequired() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { required: true }));
    }
    butNotRequired() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { required: false }));
    }
    butUnique() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { unique: true }));
    }
    butNotUnique() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { unique: false }));
    }
    withConstraints(constraints) {
        return new StringItem(Object.assign(Object.assign({}, this.description), { constraints: this.constraints.concat(constraints) }));
    }
    withoutConstraints() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { constraints: [] }));
    }
    validate(value) {
        if (typeof value !== "string") {
            return (0, _types.Left)("Only strings allowed.");
        }
        for (const c of this.constraints) {
            if (!c.requirement(value)) {
                return (0, _types.Left)(c.message);
            }
        }
        return (0, _types.Right)(value);
    }
}
exports.StringItem = StringItem;
