// userscript-metadata/dist/item.mjs
import { Left, Right } from "./types";
import { fromMaybeUndefined } from "./common";
import { not, containsLineBreaks, containsOnlyWhitespace, containsLeadingWhitespace, containsTrailingWhitespace, } from "./common";
import * as Msg from "./messages";
export class BaseItem {
    constructor(description) {
        this.description = description;
        this.key = description.key;
    }
}
export class BooleanItem extends BaseItem {
    constructor() {
        super(...arguments);
        this.unique = true;
        this.required = false;
    }
    validate(value) {
        return (value === true
            ? Right(true)
            : Left(Msg.onlyTrueAllowed));
    }
}
export class NonBooleanItem extends BaseItem {
    constructor(description) {
        super(description);
        this.description = description;
        this.required = description.required;
        this.unique = description.unique;
        this.constraints = fromMaybeUndefined([], description.constraints);
    }
}
export class StringItem extends NonBooleanItem {
    constructor(description) {
        super(description);
        this.description = description;
        this.constraints = [
            {
                requirement: not(containsOnlyWhitespace),
                message: Msg.someNonWhitespaceRequired,
            },
            {
                requirement: not(containsLeadingWhitespace),
                message: Msg.leadingWhitespaceNotAllowed,
            },
            {
                requirement: not(containsTrailingWhitespace),
                message: Msg.trailingWhitespaceNotAllowed,
            },
            {
                requirement: not(containsLineBreaks),
                message: Msg.lineBreaksNotAllowed,
            },
        ].concat(fromMaybeUndefined([], description.constraints));
    }
    butRequired() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { required: true }));
    }
    butNotRequired() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { required: false }));
    }
    butUnique() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { unique: true }));
    }
    butNotUnique() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { unique: false }));
    }
    withConstraints(constraints) {
        return new StringItem(Object.assign(Object.assign({}, this.description), { constraints: this.constraints.concat(constraints) }));
    }
    withoutConstraints() {
        return new StringItem(Object.assign(Object.assign({}, this.description), { constraints: [] }));
    }
    validate(value) {
        if (typeof value !== "string") {
            return Left("Only strings allowed.");
        }
        for (const c of this.constraints) {
            if (!c.requirement(value)) {
                return Left(c.message);
            }
        }
        return Right(value);
    }
}
