// userscript-metadata/dist/main.mjs
import { fromMaybeUndefined } from "./common";
import { fromEntries } from "./conversion";
import { Left, Right, isLeft, mapEither } from "./types";
import { stringify } from "./stringify";
import { extractBlock, parseBlock } from "./read";
import { validateEntriesWith, validateWith, UNDERSCORES_AS_HYPHENS_DEFAULT } from "./validation";
export const validateAndStringify = validateAndStringifyWith();
export function validateAndStringifyWith(options) {
    return (metadata) => {
        const validationResult = validateWith(options)(metadata);
        return mapEither(x => ({
            stringified: stringify(x.validated, options),
            warnings: x.warnings,
        }), validationResult);
    };
}
export const readAndValidate = readAndValidateWith();
export function readAndValidateWith(options = {}) {
    return (userscript) => {
        const extractResult = extractBlock(userscript);
        if (isLeft(extractResult)) {
            return Left({
                type: "EXTRACT" /* EXTRACT */,
                reason: extractResult.Left,
            });
        }
        const parseResult = parseBlock(extractResult.Right);
        if (isLeft(parseResult)) {
            return Left({
                type: "PARSE" /* PARSE */,
                lines: parseResult.Left,
            });
        }
        const validationResult = validateEntriesWith(options)(parseResult.Right);
        if (isLeft(validationResult)) {
            return Left({
                type: "VALIDATE" /* VALIDATE */,
                errors: validationResult.Left,
            });
        }
        const underscoresAsHyphens = fromMaybeUndefined(UNDERSCORES_AS_HYPHENS_DEFAULT, options.underscoresAsHyphens);
        const metadata = fromEntries(validationResult.Right.validated, underscoresAsHyphens);
        const warnings = validationResult.Right.warnings;
        return Right({ metadata, warnings });
    };
}
