// userscript-metadata/dist/main.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.readAndValidate = exports.validateAndStringify = undefined;
exports.validateAndStringifyWith = validateAndStringifyWith;
exports.readAndValidateWith = readAndValidateWith;

var _common = require("./common");

var _conversion = require("./conversion");

var _types = require("./types");

var _stringify = require("./stringify");

var _read = require("./read");

var _validation = require("./validation");

const validateAndStringify = exports.validateAndStringify = validateAndStringifyWith();
function validateAndStringifyWith(options) {
    return metadata => {
        const validationResult = (0, _validation.validateWith)(options)(metadata);
        return (0, _types.mapEither)(x => ({
            stringified: (0, _stringify.stringify)(x.validated, options),
            warnings: x.warnings
        }), validationResult);
    };
}
const readAndValidate = exports.readAndValidate = readAndValidateWith();
function readAndValidateWith(options = {}) {
    return userscript => {
        const extractResult = (0, _read.extractBlock)(userscript);
        if ((0, _types.isLeft)(extractResult)) {
            return (0, _types.Left)({
                type: "EXTRACT" /* EXTRACT */
                , reason: extractResult.Left
            });
        }
        const parseResult = (0, _read.parseBlock)(extractResult.Right);
        if ((0, _types.isLeft)(parseResult)) {
            return (0, _types.Left)({
                type: "PARSE" /* PARSE */
                , lines: parseResult.Left
            });
        }
        const validationResult = (0, _validation.validateEntriesWith)(options)(parseResult.Right);
        if ((0, _types.isLeft)(validationResult)) {
            return (0, _types.Left)({
                type: "VALIDATE" /* VALIDATE */
                , errors: validationResult.Left
            });
        }
        const underscoresAsHyphens = (0, _common.fromMaybeUndefined)(_validation.UNDERSCORES_AS_HYPHENS_DEFAULT, options.underscoresAsHyphens);
        const metadata = (0, _conversion.fromEntries)(validationResult.Right.validated, underscoresAsHyphens);
        const warnings = validationResult.Right.warnings;
        return (0, _types.Right)({ metadata, warnings });
    };
}
