// userscript-metadata/dist/messages.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.matchInsteadOfIncludeDescription = exports.matchInsteadOfIncludeSummary = exports.noMatchOrIncludeDescription = exports.noMatchOrIncludeSummary = exports.semanticVersionRequired = exports.validMatchPatternRequired = exports.trailingWhitespaceNotAllowed = exports.leadingWhitespaceNotAllowed = exports.lineBreaksNotAllowed = exports.whitespaceNotAllowed = exports.emptyStringNotAllowed = exports.someNonWhitespaceRequired = exports.onlyTrueAllowed = undefined;

var _key = require("./key");

var _syntax = require("./syntax");

const TAG_INCLUDE = (0, _syntax.tag)(_key.KEY.include);
const TAG_MATCH = (0, _syntax.tag)(_key.KEY.match);
const onlyTrueAllowed = exports.onlyTrueAllowed = "Only true allowed.";
const someNonWhitespaceRequired = exports.someNonWhitespaceRequired = "At least one non-whitespace character is required.";
const emptyStringNotAllowed = exports.emptyStringNotAllowed = "The empty string is not allowed.";
const whitespaceNotAllowed = exports.whitespaceNotAllowed = "Whitespace is not allowed.";
const lineBreaksNotAllowed = exports.lineBreaksNotAllowed = "Line breaks are not allowed.";
const leadingWhitespaceNotAllowed = exports.leadingWhitespaceNotAllowed = "Leading whitespace is not allowed.";
const trailingWhitespaceNotAllowed = exports.trailingWhitespaceNotAllowed = "Trailing whitespace is not allowed.";
const validMatchPatternRequired = exports.validMatchPatternRequired = `Value must be a valid @match pattern (e.g. "*://example.com/*").`;
const semanticVersionRequired = exports.semanticVersionRequired = `Value must follow the Semantic Versioning format, e.g. "1.0.5".`;
const noMatchOrIncludeSummary = exports.noMatchOrIncludeSummary = `No ${TAG_MATCH} or ${TAG_INCLUDE} directive.`;
const noMatchOrIncludeDescription = exports.noMatchOrIncludeDescription = `${TAG_MATCH} and/or ${TAG_INCLUDE} directives are used to indicate on which URLs the userscript should run. If neither is present, the userscript may run on ALL or NO URLs, depending on the userscript client.`;
const matchInsteadOfIncludeSummary = exports.matchInsteadOfIncludeSummary = `${TAG_INCLUDE} directive can be replaced with ${TAG_MATCH}.`;
const matchInsteadOfIncludeDescription = exports.matchInsteadOfIncludeDescription = `
The ${TAG_INCLUDE} directive is generally not recommended, since its asterisk (*) has a less safe meaning than in the ${TAG_MATCH} directive.

There is one case where ${TAG_INCLUDE} is preferable, though: It supports regular expressions, while ${TAG_MATCH} only supports match patterns as defined here:

    https://developer.chrome.com/extensions/match_patterns

Unless you need regular expressions, consider using ${TAG_MATCH} instead.
`.trim();
