// userscript-metadata/dist/patterns.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.normalizeMatchPattern = normalizeMatchPattern;
exports.isMatchPattern = isMatchPattern;
exports.isIncludePattern = isIncludePattern;
exports.isIncludePattern_regex = isIncludePattern_regex;
exports.regexFromIncludePattern = regexFromIncludePattern;
exports.schemeIn = schemeIn;
exports.hostIn = hostIn;
exports.pathIn = pathIn;

var _escapeStringRegexp = require("escape-string-regexp");

var _escapeStringRegexp2 = _interopRequireDefault(_escapeStringRegexp);

var _common = require("./common");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const REGEX_MATCH_ALL = _common.r`<all_urls>`;
const REGEX_MATCH_SCHEME = _common.r`\*|https?`;
const REGEX_MATCH_HOST = _common.r`(?:\*\.)*[^\/\*]+|\*`;
const REGEX_MATCH_PATH = _common.r`\/.*`;
const MATCH_PATTERN_ALL_NORMALIZED = _common.r`*://*/*`;
// Outer parentheses necessary to enclose `|`:
const REGEX_MATCH_PATTERN = new RegExp(_common.r`^(?:` + REGEX_MATCH_ALL + _common.r`|` + _common.r`(` + REGEX_MATCH_SCHEME + _common.r`):\/\/` + _common.r`(` + REGEX_MATCH_HOST + _common.r`)` + _common.r`(` + REGEX_MATCH_PATH + _common.r`)` + _common.r`)$`);
;
const REGEX_INCLUDE_PATTERN_SIMPLE = _common.r`^(.+)$`;
const REGEX_INCLUDE_PATTERN_REGEX = _common.r`^\/(.+)\/$`;
const REGEX_INCLUDE_PATTERN = new RegExp(REGEX_INCLUDE_PATTERN_REGEX + "|" + REGEX_INCLUDE_PATTERN_SIMPLE);
function normalizeMatchPattern(pattern) {
    return pattern === REGEX_MATCH_ALL ? MATCH_PATTERN_ALL_NORMALIZED : pattern;
}
function isMatchPattern(pattern) {
    return REGEX_MATCH_PATTERN.test(pattern);
}
function isIncludePattern(pattern) {
    return REGEX_INCLUDE_PATTERN.test(pattern);
}
function isIncludePattern_regex(pattern) {
    return new RegExp(REGEX_INCLUDE_PATTERN_REGEX).test(pattern);
}
function withoutSurroundingSlashes(s) {
    return s.replace(/^\/|\/$/g, "");
}
function regexFromIncludePattern(pattern) {
    return isIncludePattern_regex(pattern) ? new RegExp(withoutSurroundingSlashes(pattern), "i") : new RegExp(_common.r`^` + regexify(pattern) + _common.r`$`, "i");
}
function regexify(segment) {
    return (0, _escapeStringRegexp2.default)(segment).replace(new RegExp(_common.r`\\\*`, "g"), ".*");
}
// Returns null if the pattern is invalid:
function extractGroup(group, matchPattern) {
    try {
        const match = normalizeMatchPattern(matchPattern).match(REGEX_MATCH_PATTERN);
        return match === null ? null : match[group];
    } catch (_) {
        return null;
    }
}
function schemeIn(matchPattern) {
    return extractGroup(1 /* SCHEME */, matchPattern);
}
function hostIn(matchPattern) {
    return extractGroup(2 /* HOST */, matchPattern);
}
function pathIn(matchPattern) {
    return extractGroup(3 /* PATH */, matchPattern);
}
