// userscript-metadata/dist/patterns.mjs
import escapeStringRegexp from "escape-string-regexp";
import { r } from "./common";
const REGEX_MATCH_ALL = r `<all_urls>`;
const REGEX_MATCH_SCHEME = r `\*|https?`;
const REGEX_MATCH_HOST = r `(?:\*\.)*[^\/\*]+|\*`;
const REGEX_MATCH_PATH = r `\/.*`;
const MATCH_PATTERN_ALL_NORMALIZED = r `*://*/*`;
// Outer parentheses necessary to enclose `|`:
const REGEX_MATCH_PATTERN = new RegExp(r `^(?:` + REGEX_MATCH_ALL + r `|`
    + r `(` + REGEX_MATCH_SCHEME + r `):\/\/`
    + r `(` + REGEX_MATCH_HOST + r `)`
    + r `(` + REGEX_MATCH_PATH + r `)`
    + r `)$`);
;
const REGEX_INCLUDE_PATTERN_SIMPLE = r `^(.+)$`;
const REGEX_INCLUDE_PATTERN_REGEX = r `^\/(.+)\/$`;
const REGEX_INCLUDE_PATTERN = new RegExp(REGEX_INCLUDE_PATTERN_REGEX + "|" + REGEX_INCLUDE_PATTERN_SIMPLE);
export function normalizeMatchPattern(pattern) {
    return pattern === REGEX_MATCH_ALL ? MATCH_PATTERN_ALL_NORMALIZED : pattern;
}
export function isMatchPattern(pattern) {
    return REGEX_MATCH_PATTERN.test(pattern);
}
export function isIncludePattern(pattern) {
    return REGEX_INCLUDE_PATTERN.test(pattern);
}
export function isIncludePattern_regex(pattern) {
    return new RegExp(REGEX_INCLUDE_PATTERN_REGEX).test(pattern);
}
function withoutSurroundingSlashes(s) {
    return s.replace(/^\/|\/$/g, "");
}
export function regexFromIncludePattern(pattern) {
    return (isIncludePattern_regex(pattern)
        ? new RegExp(withoutSurroundingSlashes(pattern), "i")
        : new RegExp(r `^` + regexify(pattern) + r `$`, "i"));
}
function regexify(segment) {
    return escapeStringRegexp(segment).replace(new RegExp(r `\\\*`, "g"), ".*");
}
// Returns null if the pattern is invalid:
function extractGroup(group, matchPattern) {
    try {
        const match = normalizeMatchPattern(matchPattern).match(REGEX_MATCH_PATTERN);
        return match === null ? null : match[group];
    }
    catch (_) {
        return null;
    }
}
export function schemeIn(matchPattern) {
    return extractGroup(1 /* SCHEME */, matchPattern);
}
export function hostIn(matchPattern) {
    return extractGroup(2 /* HOST */, matchPattern);
}
export function pathIn(matchPattern) {
    return extractGroup(3 /* PATH */, matchPattern);
}
