// userscript-metadata/dist/stringify.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.stringify = stringify;
exports.stringifyEntries = stringifyEntries;

var _common = require("./common");

var _conversion = require("./conversion");

var _validation = require("./validation");

var _syntax = require("./syntax");

const ALIGN_DEFAULT = true;
const MINIFY_DEFAULT = false;
const SPACING_DEFAULT = 2;
function stringify(metadata, options = {}) {
    const underscoresAsHyphens = (0, _common.fromMaybeUndefined)(_validation.UNDERSCORES_AS_HYPHENS_DEFAULT, options.underscoresAsHyphens);
    const entries = (0, _conversion.toEntries)(metadata, underscoresAsHyphens);
    return stringifyEntries(entries, options);
}
function stringifyEntries(entries, options = {}) {
    // minify option overrides align and spacing.
    const minify = (0, _common.fromMaybeUndefined)(MINIFY_DEFAULT, options.minify);
    const align = minify ? false : (0, _common.fromMaybeUndefined)(ALIGN_DEFAULT, options.align);
    const spacing = minify ? 1 : (0, _common.fromMaybeUndefined)(SPACING_DEFAULT, options.spacing);
    return [].concat(_syntax.START_TAG).concat(entries.map(stringifyEntry({
        columnWidth: align ? lengthOfLongestTagIn(entries) : 0,
        spacing
    }))).concat(_syntax.END_TAG).map(line => _syntax.COMMENT_PREFIX + (minify ? "" : " ") + line).join("\n");
}
// align is basically the length of the longest tag.
function stringifyEntry(x) {
    return entry => (_syntax.TAG_PREFIX + entry.key // tag
    + " ".repeat(Math.max(0, zeroIfNaN(x.columnWidth) - entry.key.length)) // padding (at least 0)
    + " ".repeat(Math.max(1, oneIfNaN(x.spacing))) // spacing (at least 1)
    + stringifyValue(entry.value) // value
    ).trimEnd(); // no trailing whitespace on boolean tags
}
function stringifyValue(value) {
    return value === true ? "" : value;
}
function zeroIfNaN(x) {
    return Number.isNaN(x) ? 0 : x;
}
function oneIfNaN(x) {
    return Number.isNaN(x) ? 1 : x;
}
function lengthOfLongestTagIn(entries) {
    return Math.max(...entries.map(entry => entry.key.length));
}
