// userscript-metadata/dist/validation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.validate = exports.validateEntries = exports.UNDERSCORES_AS_HYPHENS_DEFAULT = undefined;
exports.validateEntriesWith = validateEntriesWith;
exports.validateWith = validateWith;

var _types = require("./types");

var _common = require("./common");

var _conversion = require("./conversion");

var _data = require("./data");

var _messages = require("./messages");

var Msg = _interopRequireWildcard(_messages);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

const UNDERSCORES_AS_HYPHENS_DEFAULT = exports.UNDERSCORES_AS_HYPHENS_DEFAULT = true;
const KEY_RULES = [{
    requirement: (0, _common.not)(_common.isEmpty),
    message: Msg.emptyStringNotAllowed
}, {
    requirement: (0, _common.not)(_common.containsWhitespace),
    message: Msg.whitespaceNotAllowed
}];
function validateKey(key) {
    for (const rule of KEY_RULES) {
        if (!rule.requirement(key)) {
            return (0, _types.Left)(rule.message);
        }
    }
    return (0, _types.Right)(key);
}
const validateEntries = exports.validateEntries = validateEntriesWith();
function validateEntriesWith(options = {}) {
    return entries => {
        const itemList = Object.values((0, _common.fromMaybeUndefined)(_data.DEFAULT_ITEMS, options.items));
        const requiredItems = itemList.filter(i => i.required);
        const duplicateItems = [];
        const errors = [];
        // Check presence of required items:
        requiredItems.forEach(item => {
            if (entries.find(entry => entry.key === item.key) === undefined) {
                errors.push({ item, kind: "REQUIRED_MISSING" /* REQUIRED_MISSING */ });
            }
        });
        // Check present entries:
        entries.forEach(entry => {
            // Validate key format first:
            const keyValidation = validateKey(entry.key);
            if ((0, _types.isLeft)(keyValidation)) {
                if (!errors.some(e => e.kind === "INVALID_KEY" /* INVALID_KEY */ && e.entry.key === entry.key)) {
                    errors.push({ entry, kind: "INVALID_KEY" /* INVALID_KEY */, reason: keyValidation.Left });
                }
            } else {
                // Key format is valid (but key may be unrecognized).
                const item = itemList.find(i => i.key === entry.key);
                if (item === undefined) {
                    if (!errors.some(e => e.kind === "UNRECOGNIZED_KEY" /* UNRECOGNIZED_KEY */ && e.entry.key === entry.key)) {
                        errors.push({ entry, kind: "UNRECOGNIZED_KEY" /* UNRECOGNIZED_KEY */ });
                    }
                } else {
                    const validation = item.validate(entry.value);
                    if ((0, _types.isLeft)(validation)) {
                        errors.push({ entry, kind: "INVALID_VALUE" /* INVALID_VALUE */, reason: validation.Left });
                    }
                    if (item.unique && entries.filter(e => e.key === entry.key).length > 1) {
                        duplicateItems.push(item);
                    }
                }
            }
        });
        duplicateItems.reduce( // to only mention each key once
        (reduced, current) => reduced.includes(current) ? reduced : reduced.concat(current), []).sort( // alphabetically
        (a, b) => a.key === b.key ? 0 : a.key < b.key ? -1 : 1).forEach(item => {
            errors.push({ item, kind: "MULTIPLE_UNIQUE" /* MULTIPLE_UNIQUE */ });
        });
        return errors.length > 0 ? (0, _types.Left)(errors) : (0, _types.Right)({
            validated: entries,
            warnings: (0, _common.fromMaybeUndefined)(_data.DEFAULT_WARNINGS, options.warnings).reduce((acc, gen) => acc.concat(gen(entries)), [])
        });
    };
}
const validate = exports.validate = validateWith();
function validateWith(options = {}) {
    const underscoresAsHyphens = (0, _common.fromMaybeUndefined)(UNDERSCORES_AS_HYPHENS_DEFAULT, options.underscoresAsHyphens);
    return metadata => {
        const entries = (0, _conversion.toEntries)(metadata, underscoresAsHyphens);
        const result = validateEntriesWith(options)(entries);
        return (0, _types.mapEither)(x => ({
            validated: metadata,
            warnings: x.warnings
        }), result);
    };
}
