// userscript-metadata/dist/validation.mjs
import { Left, Right, isLeft, mapEither } from "./types";
import { fromMaybeUndefined } from "./common";
import { toEntries } from "./conversion";
import { DEFAULT_ITEMS, DEFAULT_WARNINGS } from "./data";
import { not, isEmpty, containsWhitespace } from "./common";
import * as Msg from "./messages";
export const UNDERSCORES_AS_HYPHENS_DEFAULT = true;
const KEY_RULES = [
    {
        requirement: not(isEmpty),
        message: Msg.emptyStringNotAllowed,
    },
    {
        requirement: not(containsWhitespace),
        message: Msg.whitespaceNotAllowed,
    },
];
function validateKey(key) {
    for (const rule of KEY_RULES) {
        if (!rule.requirement(key)) {
            return Left(rule.message);
        }
    }
    return Right(key);
}
export const validateEntries = validateEntriesWith();
export function validateEntriesWith(options = {}) {
    return (entries) => {
        const itemList = Object.values(fromMaybeUndefined(DEFAULT_ITEMS, options.items));
        const requiredItems = itemList.filter(i => i.required);
        const duplicateItems = [];
        const errors = [];
        // Check presence of required items:
        requiredItems.forEach(item => {
            if (entries.find(entry => entry.key === item.key) === undefined) {
                errors.push({ item, kind: "REQUIRED_MISSING" /* REQUIRED_MISSING */ });
            }
        });
        // Check present entries:
        entries.forEach(entry => {
            // Validate key format first:
            const keyValidation = validateKey(entry.key);
            if (isLeft(keyValidation)) {
                if (!errors.some(e => e.kind === "INVALID_KEY" /* INVALID_KEY */ && e.entry.key === entry.key)) {
                    errors.push({ entry, kind: "INVALID_KEY" /* INVALID_KEY */, reason: keyValidation.Left });
                }
            }
            else {
                // Key format is valid (but key may be unrecognized).
                const item = itemList.find(i => i.key === entry.key);
                if (item === undefined) {
                    if (!errors.some(e => e.kind === "UNRECOGNIZED_KEY" /* UNRECOGNIZED_KEY */ && e.entry.key === entry.key)) {
                        errors.push({ entry, kind: "UNRECOGNIZED_KEY" /* UNRECOGNIZED_KEY */ });
                    }
                }
                else {
                    const validation = item.validate(entry.value);
                    if (isLeft(validation)) {
                        errors.push({ entry, kind: "INVALID_VALUE" /* INVALID_VALUE */, reason: validation.Left });
                    }
                    if (item.unique && entries.filter(e => e.key === entry.key).length > 1) {
                        duplicateItems.push(item);
                    }
                }
            }
        });
        duplicateItems
            .reduce(// to only mention each key once
        (reduced, current) => (reduced.includes(current) ? reduced : reduced.concat(current)), [])
            .sort(// alphabetically
        (a, b) => a.key === b.key ? 0 : a.key < b.key ? -1 : 1)
            .forEach(item => {
            errors.push({ item, kind: "MULTIPLE_UNIQUE" /* MULTIPLE_UNIQUE */ });
        });
        return (errors.length > 0
            ? Left(errors)
            : Right({
                validated: entries,
                warnings: fromMaybeUndefined(DEFAULT_WARNINGS, options.warnings).reduce((acc, gen) => acc.concat(gen(entries)), []),
            }));
    };
}
export const validate = validateWith();
export function validateWith(options = {}) {
    const underscoresAsHyphens = fromMaybeUndefined(UNDERSCORES_AS_HYPHENS_DEFAULT, options.underscoresAsHyphens);
    return (metadata) => {
        const entries = toEntries(metadata, underscoresAsHyphens);
        const result = validateEntriesWith(options)(entries);
        return mapEither(x => ({
            validated: metadata,
            warnings: x.warnings,
        }), result);
    };
}
